### DISCUSSION TITLE
[Feature] Photo editor: keep exif info, creation date and shared album

### DISCUSSION THREAD

**Discussion by german-aksenov** [immich-app/immich#11470] [Feature Request]
### I have searched the existing feature requests to make sure this is not a duplicate request.

- [X] Yes

### The feature

It nice to have in photo editor to keep original information about location, date and time, and if i rotate image in album to save 
 this rotated image in the same album.

### Platform

- [X] Server
- [ ] Web
- [X] Mobile

**Comment by juanpeich** [comment]
Do you mean to edit the photo and replace the original photo without creating duplicates?

**Comment by voitekhovich** [comment]
Immich should implement functions for editing metadata and saving it in the original file. There is a lot of archive data that does not contain metadata. Especially creating a photo or recording a video. Within Immich, editing the date works, but if you unload files, or need to reconfigure Immich from scratch, the information will be lost and you will have to configure everything again.

**Comment by SelfHostedLife** [comment]
so we don't want to be replacing the original files, that is against policy and all

how about edits in the immich editor create a new photo, with a copy of the exif and dates and all, and automatically stack it on top of the original, so that in the timeline or albums we see the edited photo?

if not the existing stack functionality, then something nearly identical, except instead of the stacking interface to see other photos, it is in the menu to "see original"? 
"remove all edits" could be in the edit menu, basically same as "revert to original"

**Reply by CristianKerr** [reply]
> so we don't want to be replacing the original files, that is against policy and all

I understand why you don't want to do that automatically but now, when I just want to rotate a photo -  I have to rotate, save (copy), go to the top of timeline, edit date, time etc to match the original, go back in timeline to the original date and delete the original photo. Somehow I feel like it would have been much better to at least ask "do you want to replace the original" and keep metadata and everything.

**Reply by lennarkivistik** [reply]
I agree with @CristianKerr on this. Every now and then I spot images that are rotated to the side, and it really breaks the flow when scrolling through the timeline. I’d love to fix them, but the current process is just too manual and risky.

For example, if I want to fix one photo, I have to:
- Open the image on the Immich phone app and rotate it
- Write down the original capture date and metadata
- Save (which creates a new file at todays date)
- Manually edit the date and metadata back in
- Re add it to the right album, Then delete the old one.

After doing that a few times, you just stop bothering. It’s not that rotation is hard, it’s that keeping everything intact (date, metadata, album) feels like doing a micro surgery.

Ideally, editing something simple like rotation should just work without breaking anything. The new version should keep the original EXIF data, stay in the same spot on the timeline, and not lose album membership. A “replace or save as new” prompt would be perfect.

**Reply by Kuphi** [reply]
Rotating an image in the app does reduce the resolution for me. From 2268 x 4032 down to 2560 x 1440 (90° rotated).

**Reply by webtweakers** [reply]
I agree with this as well.  And I've noticed that the web app does not have this EDIT function.  So I needed to go to the mobile app (Android), rotate an image, and then "save to gallery".  So Immich creates a copy and leaves it to the usual auto-backup to import the photo back into Immich, after which the date is set to today. 

Unfortunately this is extremely cumbersome, while one would expect that rotating an image is a non-destructive operation that would "simply" rotate the original image while leaving all metadata intact (date, location, people, albums, etc.).

Rotating an image is different from cropping: the latter operation IS destructive and alters the original image irreversibly. Rotating can easily be undone by rotating again, so I don't really see the need to create a copy of the image...  At least, that is assuming nothing of the original image gets lost in the rotation process (I'm referring to the previous message of @Kuphi).

**Comment by mikerjensen** [comment]
Thanks, there is [another recent request for this in a duplicate post](https://github.com/immich-app/immich/discussions/11542#discussioncomment-14980362) that was made about this issue last year, but it seems rather few users habitually do basic edits on photos, like just rotating or cropping them and I guess it's a more complex problem than it looks and may need some heavy work on the mobile app.

Google Photos assumes you just want a new edited image and leaves the user to go back and delete the original if they want, which is a bit of a pain. Not sure what Apple does. Asking to overwrite the original would be good, if there's a reliable way of doing that. I guess perhaps by writing out the new file and then changing filenames once complete.

**Comment by chienhao10** [comment]
I have tons of photos back from 20 years ago and require rotation fix... would love to rotate in immich since this is a great app that me and family members are using.

**Reply by CristianKerr** [reply]
Exactly my issue, I am importing also about 20 years worth of photos and many of them are just not properly rotated for immich.

**Reply by TokenRing** [reply]
In the end, the solution is there and easy - edit the photos how you wish OUTSIDE of immich, and then do the upload, right? Immich is easy to use from smartphones; that's the beauty of it. If you take a picture and it's not the correct orientation, you change it first, then allow it to be ingested into Immich, right?

With photo archive collections that start their life outside of modern image management systems, doesn't it make more sense to go through them first, applying the edits you want to them, and then do the upload?

Because in the end, regardless of what we think - Immich is never going to modify, to the primary developers' minds that is the same as destroying, "The Original Image" - so we have to learn to live with it.

What would be nice to see that I think would be the best compromise, is what most modern managers do: you have The Original, then as soon as you make any edits to it, the manager tool saves the original, and then each set of edits as a different, new file. It only shows the viewer the newest edit, but there's always a "relatively" easy to access button, per image, that says "revert to The Original." Sure it will consume more disk space but it obeys the cardinal rule of never losing that original file.

In summary - I sympathize but I don't think it's likely to happen soon until something like the compromise I mentioned could be implemented, and that seems like a pretty huge lift.

**Reply by webtweakers** [reply]
Not really.  I just imported a Google Photos takeout (using immich-go) and a few photos that had the correct orientation in GP are now turned sideways in Immich.  

I think image rotation should be part of an image library, especially because it doesn't affect the original (when done correctly).  Rotate the image again, and you'll have the original back.

Imagine a real-world example by throwing a bunch of photos on a desk.  Now imagine you are not allowed to rotate an image to see it correctly, because it would spontaneously create a duplicate that loses all metadata.  How awkward!

**Reply by CristianKerr** [reply]
> In the end, the solution is there and easy - edit the photos how you wish OUTSIDE of immich, and then do the upload, right? 

Bro, this is a bug report not a discussion. Telling us to do it somewhere else is like to tell a person with a broken car to walk instead... in the car repairsman office...

Just fyi image orientation is not changing the image (non-destructive) as most modern editors do this by adding a single metadata tag.

**Comment by ElternGarten** [comment]
Excuse me? You don't plan to include a rotate function? Because you think you're better than that? You deliberately don't want to do it? You're not even working on it? All those who are waiting for it have been waiting in vain?

Are you serious?

It's so important to be able to rotate a photo at exactly that point. It's no use having a new one with different information. You can't use it like that.

Why don't you create a stack? The original photo at the very back, and the rotated version in front of it. You can even put the edited version in front of that. That way you have something like a log or a history. That would be a solution that is in line with your rules. Although your rules are more important to you than the needs of your customers. -.- 

German: origian.
> Wie bitte? Ihr habt nicht vor eine drehen Funktion einzubinden? Weil ihr euch für was besseres haltet? Ihr wollt absichtlich das nicht machen? Ihr arbeitet auch nicht dran? Alle die darauf warten, waren vergeblich? Ist das euer ernst? Es ist so wichtig, dass genau an der Stelle ein gedrehtes Foto möglich sein wird. Es bringt nichts ein neues mit anderen Informationen zu haben. So kann man das nicht benutzen. Warum ersteltl irh nicht einen Stapel? Das original Foto ganz hinten, und davor die gedrehte Version. Davor kann man sogar die beabrietet Version noch stellen. So hat man sowas wie ein Protokoll oder eine Hystorie. Das wäre doch eien Lsöugn, die im einklag mit eurem Regelwerk steht. Wobei euer Regelwerk euch wichtigger als die Kunden bedürfnisse ist. -.-

**Reply by CristianKerr** [reply]
Buddy calm down. The rotate function is there, the issue is it creates copy of the photo without all exifdata so its not an easy rotate and save task. We have to rotate, save, add metadata, delete old photo and its quite painful.

I agree with you that this should be like one of the basic functions but I don't agree with the tone, you don't need to be toxic about it.

**Comment by mikerjensen** [comment]
I agree that an image can be altered significantly during
      cropping, but I see no reason not to still retain EXIF information
      regarding the location and the camera used to produce the image.
    Mike
    On 1/5/26 11:22, Bas Van Gaalen wrote:
    
    
      
      I agree with this as well. And I've noticed that the
        web app does not have this EDIT function. So I needed to go to
        the mobile app (Android), rotate an image, and then "save to
        gallery". So Immich creates a copy and leaves it to the usual
        auto-backup to import the photo back into Immich, after which
        the date is set to today.
      Unfortunately this is extremely cumbersome, while
        one would expect that rotating an image is a non-destructive
        operation that would "simply" rotate the original image while
        leaving all metadata intact (date, location, people, albums,
        etc.).
      Rotating an image is different from cropping: the
        latter operation IS destructive and alters the original image
        irreversibly. Rotating can easily be undone by rotating again,
        so I don't really see the need to create a copy of the image...
        At least, that is assuming nothing of the original image gets
        lost in the rotation process (I'm referring to the previous
        message of @Kuphi).
      —
        Reply to this email directly, view it on GitHub, or unsubscribe.
        You are receiving this because you commented.Message
          ID: ***@***.***>
      [
{
***@***.***": "http://schema.org",
***@***.***": "EmailMessage",
"potentialAction": {
***@***.***": "ViewAction",
"target": "https://github.com/immich-app/immich/discussions/11470#discussioncomment-15415074",
"url": "https://github.com/immich-app/immich/discussions/11470#discussioncomment-15415074",
"name": "View Discussion"
},
"description": "View this Discussion on GitHub",
"publisher": {
***@***.***": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

**Comment by iMonZ** [comment]
Since this issue is pretty major, is it possible to put a warning inside the app? Its not a good solution to wipe the exif data without the users permission.

**Reply by cptkng** [reply]
The exif data of the original photo is not wiped. I think this statement is a bit misleading. But the copy of the "edited" photo will not have 'correct/expected' exif data.

**Reply by iMonZ** [reply]
> The exif data of the original photo is not wiped. I think this statement is a bit misleading. But the copy of the "edited" photo will not have 'correct/expected' exif data.

Yes but this is not stated. It would be great to get a warning if you would alter the original picture. So not only tech savvy people could use immich. The way it’s currently implemented a normal user would not know this and would maybe delete the original picture without getting any warnings for the consequences. Since this behavior of immich is not normal. You can make a checkmark at the warning to never show it again. But even when you edit a picture and the picture quality get lost a warning would be necessary since you want to trust the tool that backups your precious photo library.
Maybe a approach like Apple is doing with the .aae files would solve all of those issues permanently, since this approach never edits the original file except if you export it.
https://wphtaccess.com/2025/08/19/what-is-an-aae-file-on-iphone-can-you-safely-delete-it/

**Comment by cptkng** [comment]
I also wanted to chime in here and add my 2 cents. 

For me the behavior was also very confusing. I edited a rather old photo in my photo stream, actually I just rotated it. Expecting that this very photo showed up rotated now where it was (so from the date perspective), at a first look nothing had changed. The photo was still not rotated. I closed the app, reopened, logged out and in assuming this was a caching issue, but the photo still showed up unchanged.
A day or so later when looking at the recent photos in my timeline I finally understood what has happened. There on top with the timestamp of the edit operation a rotated copy of my photo showed up...

This is quite frustrating. 

Unrelated to how Google Photos might handle it or how others expect exif data on edit operations. My stance on this is:
Any edit operation of a photo should retain the original exif data that is relevant for immich, so creation time, location, camera and settings...
Why? 
I think for rotations this is quite clear. If you had a physical photo in a physical album, and you just 'rotate' it there, it's still the same photo. No person would now say: "ah the new creation time is set to now, where I rotated the photo", or you would not make a copy of that very photo and rotate that by 90°.
But also for all other 'normal' edit operations (filters, crop, rotate non-rectangular, ...) my expectations is that exif data is preserved. The scenery of that photo was still taken on the original data and location and with the original camera,...

I think the only other functionalty for users should be to ask them if they want to create a copy when saving the edit. 

But for now I think it should be. 
- warn a user that the edit operation changes the original (if confirmation dialogue is taking too long) - don't create a copy
- retain all exif info

Or
 - keep the current behavior with the copy. But also copy exif info. (Question is how to handle duplicate detection for minor edits, probably we need to exclude minor edits from duplicate detection to not accidentally delete the original during bulk duplicate deletion actions)





PS: Sorry if that sounds too negative, I really love this app/server component, it's simply awesome, it's so awesome that I could convince others to use it instead of Google Photos ;)

**Reply by chienhao10** [reply]
Just my opinion, warning user would be very annoying when i am editing over hundreds of photos, even the photo has been edit, as long as exif data is there then there is nothing to worry about. I tried editing the project a little and i was able to complete save the exif info, and rotate the photo. The photo itself has been "edited", but the exif info is still there i would consider the original photo and the new photo are the same.

**Reply by iMonZ** [reply]
> Just my opinion, warning user would be very annoying when i am editing over hundreds of photos, even the photo has been edit, as long as exif data is there then there is nothing to worry about. I tried editing the project a little and i was able to complete save the exif info, and rotate the photo. The photo itself has been "edited", but the exif info is still there i would consider the original photo and the new photo are the same.

You could show the warning with an optional checkbox to never show this warning again. Especially for not so tech savvy people I think this is required.

**Reply by webtweakers** [reply]
Interesting how you bring up the physical equivalent independently.  My 2 cents is that rotating an image is not _really_ an edit.  On a technical level: sure, but on the level of any ordinary non tech-savvy user: no.  Maybe there should simply be a button on the photo viewing page that rotates the image (and thumb).

**Reply by iMonZ** [reply]
> Interesting how you bring up the physical equivalent independently. My 2 cents is that rotating an image is not _really_ an edit. On a technical level: sure, but on the level of any ordinary non tech-savvy user: no. Maybe there should simply be a button on the photo viewing page that rotates the image (and thumb).

If the file is altered it’s an edit. That’s the whole problem here.

**Reply by webtweakers** [reply]
I would argue that there are different types of edits: destructive edits and non-destructive edits.  Rotating an image does not destroy the image.

**Reply by CristianKerr** [reply]
> > Interesting how you bring up the physical equivalent independently. My 2 cents is that rotating an image is not _really_ an edit. On a technical level: sure, but on the level of any ordinary non tech-savvy user: no. Maybe there should simply be a button on the photo viewing page that rotates the image (and thumb).
> 
> If the file is altered it’s an edit. That’s the whole problem here.

Actually, rotating an image is usually done by adding metadata to it. In particular an Orientation exif tag. So it is weird that this is considered an edit and it results in not only a new image file, but a one with removed metadata...

**Comment by cptkng** [comment]
Seems some of the requested features will be added soon, first for the webapp, later for mobile: https://github.com/immich-app/immich/pull/24155

**Comment by bo0tzz** [comment]
#24155
